prompter-source.lisp

```diff
@@ -790,18 +724,18 @@ If SIZE is NIL, capacity is infinite."
              (lambda (slot)
                (list (intern (symbol-name slot) "KEYWORD")
                      (slot-value object slot)))
-             (mopu:slot-names class-sym))))))
-
-(defgeneric destroy (source)
-  (:method ((source source))
-    ;; Ignore errors in case thread is already terminated.
-    ;; REVIEW: Is there a cleaner way to do this?
-    (ignore-errors (bt:destroy-thread (update-thread source)))
-    (ignore-errors (bt:destroy-thread (attribute-thread source))))
-  (:documentation "Clean up the source.
-SOURCE should not be used once this has been run."))
-
-(defun update (source input new-ready-channel) ; TODO: Store `input' in the source?
+             (slot-names class-sym))))))
+
+(define-generic destroy ((source source))
+  "Clean up the source.
+SOURCE should not be used once this has been run."
+  (maybe-funcall (destructor source) source)
+  (alex:when-let ((lpara:*kernel* (kernel source)))
+    (lpara:kill-tasks :default)
+    (lpara:end-kernel)
+    (setf (kernel source) nil)))
+
+(defun update (source input)            ; TODO: Store `input' in the source?
```

Comment:
In the past I remember that I really wanted to access `input` from source. It is a bit redundant, since `input` can be fetched from the prompter that owns the source. The issue is that, when operating on the level of source, you can't access to the prompter, since sources don't inherit from prompter objects.

In short, I'd recommend storing input in sources.

Reply:
Actually now with this PR `source`s have a `prompter` slot!

It would only be useful for stand-alone `source` objects, which rarely occurs in practice.
Any good use case in mind?

Reply:
> Actually now with this PR sources have a prompter slot!

Oh, of course! 

> It would only be useful for stand-alone source objects, which rarely occurs in practice.
Any good use case in mind?

None...